docs: add Forgejo reverse-proxy SSO guide with runnable example (#2220)
pomerium/documentation: content/docs/guides/forgejo.mdx

lang: en-US
keywords:
  [
    pomerium,
    identity access proxy,
    forgejo,
    gitea,
    git,
    reverse proxy authentication,
    single sign-on,
    sso,
    identity headers,
  ]
description: This guide shows you how to put a Forgejo instance behind Pomerium and use reverse-proxy authentication so users sign in once and are auto-provisioned from their Pomerium identity.
---

import ForgejoDockerCompose from '@site/content/examples/forgejo/forgejo.docker-compose.md';
import ForgejoPomeriumConfig from '@site/content/examples/forgejo/forgejo.pomerium-config.md';

# Secure Forgejo with Pomerium

Learn how to add authentication, authorization, and single sign-on to a [Forgejo](https://forgejo.org/) instance with Pomerium. The same reverse-proxy authentication pattern applies to [Gitea](https://about.gitea.com/), which Forgejo is a soft-fork of, but this guide uses the Forgejo container image and `FORGEJO__` environment variables.

## Background

Forgejo is a self-hosted Git service. Like most Git servers, it has its own local username and password database. Without single sign-on, every user has a second set of credentials in addition to your identity provider.

> Forgejo also supports [reverse-proxy authentication](https://forgejo.org/docs/latest/admin/setup/reverse-proxy/): it trusts an identity supplied in an HTTP header by a proxy in front of it. Pomerium is exactly that proxy. After Pomerium authenticates a user against your identity provider, it forwards the user's identity as request headers; Forgejo reads those headers, signs the user in, and (optionally) auto-creates the account on first visit. The result is single sign-on for the Forgejo **web UI** with no second password.

Reviewer: nit: the phrase "which Forgejo is a soft-fork of" feels awkward to me.

I have a few other suggestions for tweaking the flow here, feel free to take or leave.

```suggestion
Learn how to add authentication, authorization, and single sign-on to a [Forgejo](https://forgejo.org/) instance using Pomerium. Forgejo is a self-hosted Git service with a web UI.

A similar setup can also be used with [Gitea](https://about.gitea.com/).

## Background

Like many applications, Forgejo has its own local username and password database. Without single sign-on, every user would need to maintain a second set of credentials.

However, Forgejo also supports [reverse-proxy authentication](https://forgejo.org/docs/latest/admin/setup/reverse-proxy/): it trusts an identity supplied in an HTTP header by a proxy in front of it. Pomerium is exactly that proxy. After Pomerium authenticates a user against your identity provider, it forwards the user's identity in request headers. Forgejo reads these headers, signs the user in, and can even automatically create an account on first visit. The result is single sign-on for the Forgejo web UI with no second password.
```

- **Forgejo** trusts those headers (reverse-proxy authentication) and auto-registers users from them.

:::danger Forgejo must be reachable only through Pomerium

Reverse-proxy authentication means Forgejo believes the identity headers it receives. Forgejo does not verify Pomerium's signed `X-Pomerium-Jwt-Assertion` for login, so this integration depends entirely on **network isolation**: Forgejo must be reachable **only** through Pomerium. `REVERSE_PROXY_TRUSTED_PROXIES` does not gate this; it only controls how Forgejo reads `X-Forwarded-For`. Reverse-proxy auth trusts the identity header from any peer that can reach Forgejo, so the network path is the entire boundary.

If another client can connect to Forgejo directly, it can set the same identity headers and impersonate a user. In Docker Compose, do not publish Forgejo's HTTP port and keep only Pomerium on its network. In Kubernetes, a `ClusterIP` service alone is not enough if arbitrary pods can connect to Forgejo; use a policy-enforcing CNI (container network interface) and a `NetworkPolicy`, or another network control, to restrict the backend to Pomerium.

> When an upstream app can verify the signed Pomerium JWT directly, prefer that model. Forgejo does not support that login mode today, so this guide uses header-based reverse-proxy authentication with a narrow trust boundary.

Reviewer: I'd suggest trimming this a bit.

```suggestion
Reverse-proxy authentication means Forgejo believes the identity headers it receives. Forgejo does not verify Pomerium's signed `X-Pomerium-Jwt-Assertion` for login, so this integration depends entirely on **network isolation**: Forgejo must be reachable **only** through Pomerium. If another client can connect to Forgejo directly, it can set the same identity headers and impersonate a user.

In Docker Compose, do not publish Forgejo's HTTP port and keep only Pomerium on its network. In Kubernetes, a `ClusterIP` service alone is not enough if arbitrary pods can connect to Forgejo; use a policy-enforcing CNI (container network interface) and a `NetworkPolicy`, or another network control, to restrict the backend to Pomerium.

Whenever possible, prefer using Pomerium JWT authentication, which does not depend on network isolation. Forgejo does not support that login mode today, so this guide uses header-based reverse-proxy authentication.
```

  --username recovery-admin --email recovery-admin@example.com \
  --random-password --admin
```

To sign in with that local account you must reach Forgejo without Pomerium's injected header, for example on a trusted network or by temporarily setting `ENABLE_REVERSE_PROXY_AUTHENTICATION` to `false`, since normal traffic always arrives pre-authenticated through Pomerium.

## Troubleshooting

- **Pomerium returns `403`.** The route policy still allows only `user@example.com`. Replace it with your user or domain.
- **You reach Forgejo but are not signed in or auto-provisioned.** The header names in `config.yaml` (`jwt_claims_headers`) and in Forgejo (`REVERSE_PROXY_AUTHENTICATION_USER`/`_EMAIL`/`_FULL_NAME`) must match exactly, and `pass_identity_headers` must be set on the route.
- **Sign-in fails on the username.** Forgejo rejects usernames containing characters such as `@`. Use a stable, already-valid claim for the username (this guide uses `sub`).
- **A direct request to the backend signs in.** Expected: reverse-proxy auth trusts the header from any peer. It is why Forgejo must stay reachable only through Pomerium.

## Clean up

> Stop the stack with `docker compose down`. Add `-v` only to also delete the `forgejo-data` volume and all Forgejo state. To undo the integration while keeping data, remove the Forgejo route from `config.yaml` and unset the `ENABLE_REVERSE_PROXY_*` settings, after creating a break-glass admin as above.

Reviewer: nit: "only to also delete" feels awkward to me, maybe something like this instead?

```suggestion
Stop the stack with `docker compose down`. To delete the `forgejo-data` volume (with all Forgejo state), use `docker compose down -v`. To undo the integration while keeping data, remove the Forgejo route from `config.yaml` and unset the `ENABLE_REVERSE_PROXY_*` settings, after creating a break-glass admin as above.
```
